benchmark/test/test_script.py

```diff
@@ -1,10 +1,124 @@
 # Copyright 2015 ClusterHQ Inc.  See LICENSE file for details.
 
+from contextlib import contextmanager
+from cStringIO import StringIO
+import os
+import shutil
+import sys
+import tempfile
+
 from jsonschema.exceptions import ValidationError
 
 from twisted.trial.unittest import SynchronousTestCase
 
-from benchmark.script import validate_configuration, get_config_by_name
+from benchmark.script import (
+    BenchmarkOptions, get_cluster, validate_configuration, get_config_by_name
+)
+
+
+@contextmanager
+def capture_stderr():
+    """
+    Context manager to capture stderr.
+
+    Call the returned context variable to obtain captured output.
+    """
+    s = StringIO()
+    out, sys.stderr = sys.stderr, s
+    yield s.getvalue
+    sys.stderr = out
+
+# Addresses must be different, to check that environment is not used
+# during YAML tests.
+_ENV_CONTROL_SERVICE_ADDRESS = '10.0.0.1'
+_YAML_CONTROL_SERVICE_ADDRESS = '10.1.0.1'
+
+# Uses % formatting to prevent confusion with YAML braces.
+_CLUSTER_YAML_CONTENTS = '''
+control_node: %s
+agent_nodes: []
+''' % _YAML_CONTROL_SERVICE_ADDRESS
+
+
+class ClusterConfigurationTests(SynchronousTestCase):
+    """
+    Tests for mapping cluster configuration to cluster.
+    """
+
+    def setUp(self):
+        self.environ = {
+            'FLOCKER_ACCEPTANCE_DEFAULT_VOLUME_SIZE': '107374182400',
+            'FLOCKER_ACCEPTANCE_CONTROL_NODE': _ENV_CONTROL_SERVICE_ADDRESS,
+            'FLOCKER_ACCEPTANCE_HOSTNAME_TO_PUBLIC_ADDRESS': '{}',
+            'FLOCKER_ACCEPTANCE_VOLUME_BACKEND': 'openstack',
+            'FLOCKER_ACCEPTANCE_TEST_VOLUME_BACKEND_CONFIG':
+                '/tmp/tmp84DVr3/dataset-backend.yml',
+            'FLOCKER_ACCEPTANCE_NUM_AGENT_NODES': '2',
+            'FLOCKER_ACCEPTANCE_API_CERTIFICATES_PATH': '/tmp/tmpSvE7ug',
+        }
+
+    def test_environment_setup(self):
+        """
+        Uses environment variables for cluster configuration if option missing.
+        """
+        options = BenchmarkOptions()
+        options.parseOptions([])
+        cluster = get_cluster(options, self.environ)
+        self.assertEqual(
+            cluster.control_node_address().exploded,
+            _ENV_CONTROL_SERVICE_ADDRESS
+        )
+
+    def test_yaml_setup(self):
+        """
+        Uses YAML file for cluster configuration if option given.
+
+        This is true even if the environment contains a valid configuration.
+        """
+        tmpdir = tempfile.mkdtemp()
+        try:
+            with open(os.path.join(tmpdir, 'cluster.yml'), 'wt') as f:
+                f.write(_CLUSTER_YAML_CONTENTS)
+            options = BenchmarkOptions()
+            options.parseOptions(['--cluster', tmpdir])
+            cluster = get_cluster(options, self.environ)
+        finally:
+            shutil.rmtree(tmpdir)
```

Comment:
D'oh! I looked for a context manager to do this cleanly, but it only exists in Python 3. And that made me miss the `addCleanup` solution. Does Python version numbering refer to the number of obvious ways to do it?